samcli/lib/utils/git_repo.py

```diff
@@ -0,0 +1,180 @@
+
+    Attributes
+    ----------
+    _url: str
+        The URL of this Git repository, example "https://github.com/aws/aws-sam-cli"
+    _name: str
+        The name of this Git repository, example "aws-sam-cli"
+    _local_path: Path
+        The path of the local clone of this Git repository
+    _clone_attempted: bool
+        whether an attempt to clone this Git repository took place or not
+
+    Methods
+    -------
+    clone(self, clone_dir: Path, clone_name=None, replace_existing=False) -> Path:
+        creates a local clone of this Git repository
+    """
+
+    def __init__(self, url: str, name: str) -> None:
+        self._url: str = url
+        self._name: str = name
+        self._local_path: Optional[Path] = None
+        self._clone_attempted: bool = False
+
+    @property
+    def url(self) -> str:
+        return self._url
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def local_path(self) -> Optional[Path]:
+        return self._local_path
+
+    @local_path.setter
+    def local_path(self, value: Path) -> None:
+        self._local_path = value
+
+    @property
+    def clone_attempted(self) -> bool:
+        return self._clone_attempted
+
+    @clone_attempted.setter
+    def clone_attempted(self, value: bool) -> None:
+        self._clone_attempted = value
+
+    @staticmethod
+    def _ensure_clone_directory_exists(clone_dir: Path) -> None:
+        try:
+            clone_dir.mkdir(mode=0o700, parents=True, exist_ok=True)
+        except OSError as ex:
+            LOG.warning("WARN: Unable to create clone directory.", exc_info=ex)
+            raise
+
+    @staticmethod
+    def _git_executable() -> str:
+        if platform.system().lower() == "windows":
+            executables = ["git", "git.cmd", "git.exe", "git.bat"]
+        else:
+            executables = ["git"]
+
+        for executable in executables:
+            try:
+                subprocess.Popen([executable], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+                # No exception. Let's pick this
+                return executable
+            except OSError as ex:
+                LOG.debug("Unable to find executable %s", executable, exc_info=ex)
+
+        raise OSError("Cannot find git, was looking at executables: {}".format(executables))
+
+    def clone(self, clone_dir: Path, clone_name: Optional[str] = None, replace_existing: bool = False) -> Path:
+        """
+        creates a local clone of this Git repository
+
+        Parameters
+        ----------
+        clone_dir: Path
+            The directory to create the local clone inside
+        clone_name: Optional[str]
+            The dirname of the local clone, if not provided, repository's name (self._name) will be used
+        replace_existing: bool
+            Whether to replace the current local clone directory if already exists or not
+
+        Returns
+        -------
+            The path of the created local clone
```

Comment:
from the usage, it does not need to return anything from this method

Reply:
It make sense to return this path. 

why in `InitTemplates` we read from `local_path` instead? because the behavior of `InitTemplates` is to not attempt clone if already did before

but generally:
```
local_clone = git_repo.clone(....)
```

is better than
```
git_repo.clone(...)
local_clone = git_repo.local_path
```

Reply:
If that's the case, why we need to have `git_repo.local_path`?